src/main/java/uk/gov/di/ipv/cri/common/library/service/SessionService.java

```diff
@@ -84,16 +89,69 @@ public SessionItem getSession(String sessionId) {
-                        SessionItem.ACCESS_TOKEN_INDEX, accessToken.toAuthorizationHeader()));
+    public SessionItem getSessionByAccessToken(AccessToken accessToken)
+            throws SessionExpiredException, AccessTokenExpiredException, SessionNotFoundException {
+        SessionItem sessionItem;
+
+        try {
+            sessionItem =
+                    listUtil.getOneItemOrThrowError(
+                            dataStore.getItemByIndex(
+                                    SessionItem.ACCESS_TOKEN_INDEX,
+                                    accessToken.toAuthorizationHeader()));
+        } catch (IllegalArgumentException e) {
+            if (e.getMessage().contains("No items found")) {
+                throw new SessionNotFoundException("no session found with that access token");
+            } else {
+                throw new SessionNotFoundException(
+                        "more than one session found with that access token");
+            }
+        }
+
+        // Re-fetch our session directly to avoid problems with projections
+        sessionItem = getSession(String.valueOf(sessionItem.getSessionId()));
+
+        if (sessionItem.getExpiryDate() < clock.instant().getEpochSecond()) {
+            throw new SessionExpiredException("session expired");
+        }
+
+        if (sessionItem.getAccessTokenExpiryDate() < clock.instant().getEpochSecond()) {
+            throw new AccessTokenExpiredException("access code expired");
+        }
+
+        return sessionItem;
     }
 
-    public SessionItem getSessionByAuthorisationCode(String authCode) {
-        SessionItem sessionItem =
-                listUtil.getOneItemOrThrowError(
-                        dataStore.getItemByIndex(SessionItem.AUTHORIZATION_CODE_INDEX, authCode));
-        return getSession(sessionItem.getSessionId().toString());
+    public SessionItem getSessionByAuthorisationCode(String authCode)
+            throws SessionExpiredException, AuthorizationCodeExpiredException,
+                    SessionNotFoundException {
+        SessionItem sessionItem;
+
+        try {
+            sessionItem =
+                    listUtil.getOneItemOrThrowError(
+                            dataStore.getItemByIndex(
+                                    SessionItem.AUTHORIZATION_CODE_INDEX, authCode));
+        } catch (IllegalArgumentException e) {
+            if (e.getMessage().contains("No items found")) {
+                throw new SessionNotFoundException("no session found with that authorization code");
+            } else {
+                throw new SessionNotFoundException(
+                        "more than one session found with that authorization code");
+            }
+        }
+
+        // Re-fetch our session directly to avoid problems with projections
+        sessionItem = getSession(String.valueOf(sessionItem.getSessionId()));
+
+        if (sessionItem.getExpiryDate() < clock.instant().getEpochSecond()) {
```

Comment:
avoid duplicating the checking of the session expiry date and auth code expiry date & introduce a new method to perform these checks